test/Microsoft.AspNetCore.SignalR.Redis.Tests/Docker.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.SignalR.Redis.Tests
+{
+    public class Docker
+    {
+        private static readonly string _exeSuffix = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? ".exe" : string.Empty;
+
+        private static readonly string _dockerContainerName = "redisTestContainer";
+        private static Lazy<Docker> _instance = new Lazy<Docker>(Create);
+
+        public static Docker Default => _instance.Value;
+
+        private string _path;
+        public Docker(string path)
+        {
+            _path = path;
+        }
+
+        private static Docker Create()
+        {
+            var location = GetDockerLocation();
+            return location == null ? null : new Docker(location);
+        }
+
+        private static string GetDockerLocation()
+        {
+            foreach (var dir in Environment.GetEnvironmentVariable("PATH").Split(Path.PathSeparator))
+            {
+                var candidate = Path.Combine(dir, "docker" + _exeSuffix);
+                if (File.Exists(candidate))
+                {
+                    return candidate;
+                }
+            }
+
+            return null;
+        }
+
+        public int Start(ILogger logger)
+        {
+            logger.LogInformation("Starting docker container");
+
+            // create and run docker container, remove automatically when stopped, map 6379 from the container to 6379 localhost
+            // use static name 'redisTestContainer' so if the container doesn't get removed we don't keep adding more
+            // use redis base docker image
+            return RunProcess(_path, $"run --rm -p 6379:6379 --name {_dockerContainerName} -d redis", logger);
```

Comment:
Can these tests run in parallel or will they all be synchronous? Maybe pass in the port as a param to start.

Reply:
Yes this is pretty damning I think.

Reply:
This is run from the server fixture and so will only run once before the tests that use it start.